### ISSUE TITLE
Unpacking concatenated data

### ISSUE THREAD

**Issue by vovanec** [mcollina/msgpack5#9]
Hey Matteo,

I've just run into an issue trying to decode multiple msgpack strings concatenated. Assume the following sample:

```
var pack = require('msgpack5')(),
    data    = [
        { hello: 1 },
        { hello: 2 },
        { hello: 3 }
    ],
    buffers = [];

data.forEach(function (elem) {
    buffers.push(pack.encode(elem));
});   <<<--- here I have three msgpack strings concatenated. I assume I can decode them easily, iterating over the buffer one by one...

console.log(pack.decode(Buffer.concat(buffers)));
console.log(pack.decode(Buffer.concat(buffers)));
console.log(pack.decode(Buffer.concat(buffers)));
```

I have the output like:

```
{ hello: 1 }
{ hello: 1 }
{ hello: 1 }
```

My question - is there any method, to have the whole stream to be unpacked using msgpack5 library?

Thanks,
--Vovan

**Comment by msealand** [comment]
Could you use the `decoder()` stream functionality to do what you need?

Something like:

```
var pack = require('msgpack5')(),
    data    = [
        { hello: 1 },
        { hello: 2 },
        { hello: 3 }
    ];

var decoder = pack.decoder({header: false});
decoder.on('data', function(msg) {
    console.log(msg);
});

data.forEach(function (elem) {
    decoder.write(pack.encode(elem));
});
```

Outputs:

```
{ hello: 1 }
{ hello: 2 }
{ hello: 3 }
```

**Comment by mcollina** [comment]
TLDR yes, you are right

Sorry about not replying early, the response was left in the draft of my inbox, and never sent :(.

Here it is:

Your code cannot work because you are not consuming the Buffer (you cannot shorten a Buffer). You can do that by passing always the same instance of http://npm.im/bl instead of Buffer.concat, which is what msgpack5 uses internally to track consumption.

As for streaming, parsing msgpack on a streaming way is not efficient, as it does not include how long is a going to be a message. In order to support this feature, it's better to use an header to define how long is the message, so we know when a message is ready to be parsed. This was already done for you in https://github.com/mcollina/msgpack5/blob/master/README.md#encoder and https://github.com/mcollina/msgpack5/blob/master/README.md#decoder. If you see other solutions, please open another issue or send a PR!

BTW naming and docs might not be perfect. If you think so, just send a PR and we'll discuss there.

**Comment by mcollina** [comment]
TLDR yes, you are right

Sorry about not replying early, the response was left in the draft of my inbox, and never sent :(.

Here it is:

Your code cannot work because you are not consuming the Buffer (you cannot shorten a Buffer). You can do that by passing always the same instance of http://npm.im/bl instead of Buffer.concat, which is what msgpack5 uses internally to track consumption.

As for streaming, parsing msgpack on a streaming way is not efficient, as it does not include how long is a going to be a message. In order to support this feature, it's better to use an header to define how long is the message, so we know when a message is ready to be parsed. This was already done for you in https://github.com/mcollina/msgpack5/blob/master/README.md#encoder and https://github.com/mcollina/msgpack5/blob/master/README.md#decoder. If you see other solutions, please open another issue or send a PR!

BTW naming and docs might not be perfect. If you think so, just send a PR and we'll discuss there.

**Comment by vovanec** [comment]
Well, here's the real use case from my current project. Our Python web
server sends msgpack messages in format like:

ext32(data) where data is msgpack(session_id) +
msgpack(some_internal_payload)

I'm receiving this message in a client via the websocket so it's not
bufferlist but rather string or Buffer. Current work around, after external
ext32 is unpacked, is to split data onto encoded session_id and
internal_payload, then decode one by one. I can do this, but only if I know
encoded session_id length, that's not always possible as session_id has
variable length.

With python msgpack I can easily decode concatenated msgpack objects(like
msgpack(str1) + msgpack(str2) + ... ) by creating decoder object and
iterating over it. Is there anything possible with msgpack5?

Thanks,
--Vovan


2015-01-29 13:38 GMT-08:00 Matteo Collina <notifications@github.com>:

**Comment by vovanec** [comment]
I see my first example was not quite correct. Here's what I was meaning:

```javascript
var pack = require('msgpack5')(),
    data    = [
        { hello: 1 },
        { hello: 2 },
        { hello: 3 }
    ],
    buffers = [],
    decoder = pack.decoder();

decoder.on('data', function (data) {
    console.log(data);
});

data.forEach(function (elem) {
    buffers.push(pack.encode(elem));
});

var concatenated = Buffer.concat(buffers);

decoder.write(concatenated);
```

**Comment by vovanec** [comment]
It looks like this is related to https://github.com/mcollina/msgpack5/pull/2 that is already closed. If this is already fixed - can you point me how do I unpack concatenated buffers?

**Comment by mcollina** [comment]
```
var pack = require('./')()
var bl = require('bl')
var buf = bl([pack.encode('hello'), pack.encode('world')])

console.log(pack.decode(buf))
console.log(pack.decode(buf))
```

Not the best API, but still..

**Comment by mcollina** [comment]
Closing now, let me know if you want to submit a PR to improve the behavior!
